src/core/net/dns_headers.cpp

```diff
@@ -798,38 +798,103 @@ otError ResourceRecord::ReadFrom(const Message &aMessage, uint16_t aOffset)
     return error;
 }
 
+void TxtEntry::Iterator::Init(const uint8_t *aTxtData, uint16_t aTxtDataLength)
+{
+    SetTxtData(aTxtData);
+    SetTxtDataLength(aTxtDataLength);
+    SetTxtDataPoistion(0);
+}
+
+otError TxtEntry::Iterator::GetNextEntry(TxtEntry &aEntry)
+{
+    otError     error = OT_ERROR_NONE;
+    uint8_t     length;
+    uint8_t     index;
+    const char *cur;
+    char *      keyBuffer = GetKeyBuffer();
+
+    static_assert(sizeof(mChar) == TxtEntry::kMaxKeyLength + 1, "KeyBuffer cannot fit the max key length");
+
+    VerifyOrExit(GetTxtData() != nullptr, error = OT_ERROR_PARSE);
+
+    aEntry.mKey = keyBuffer;
+
+    while ((cur = GetTxtData() + GetTxtDataPosition()) < GetTxtDataEnd())
+    {
+        length = static_cast<uint8_t>(*cur);
+
+        cur++;
+        VerifyOrExit(cur + length <= GetTxtDataEnd(), error = OT_ERROR_PARSE);
+        IncreaseTxtDataPosition(sizeof(uint8_t) + length);
+
+        // Silently skip over an empty string or if the string starts with
+        // a `=` character (i.e., missing key) - RFC 6762 - section 6.4.
+
+        if ((length == 0) || (cur[0] == kKeyValueSeparator))
+        {
+            continue;
+        }
+
+        for (index = 0; index < length; index++)
+        {
+            if (cur[index] == kKeyValueSeparator)
+            {
+                keyBuffer[index++]  = kNullChar; // Increment index to skip over `=`.
+                aEntry.mValue       = reinterpret_cast<const uint8_t *>(&cur[index]);
+                aEntry.mValueLength = length - index;
+                ExitNow();
+            }
+
+            VerifyOrExit(index < kMaxKeyLength, error = OT_ERROR_PARSE);
```

Comment:
Should we use a different error here? RFC 6763 doesn't strictly forbid longer key lengths. This error should really indicate that this is a limitation of this implementation, not a problem with the TXT record itself.

Reply:
Good point.
Changed it to return `OT_ERROR_NO_BUFS` (updated the `@retval` docs to mention what this error represents).

Reply:
Thinking about this more, I wonder if it might be useful to allow the user to skip over this "too long" key? Maybe they are looking for a key that is not too long. Thoughts?

Reply:
I agree. It makes sense to let user skip over long keys.

I had this idea originally of returning things that we can not parse successfully in the `otDnsTxtEntry` with `mKey` set to `nullptr` and then return the encoded value in `mValue` and `mValueLength`. This fits well with how user is expected to provide the encoded txt entry to us.  I did not add it because I thought we cannot assume in general case that all an encoded TXT entry starts with a `length` byte (I think it can be anything in general case).

But in the case of long keys (assuming  DNS-SD format), I think using this approach fits well. I will change the code to use this model for the long key case (i.e. return success with `mKey` set to null and entire value given back to user).

Reply:
Added this behavior in the latest pushed commit (also updated unit test to cover the long key case).